src/libcore/str/mod.rs

```diff
@@ -1156,8 +1155,42 @@ fn run_utf8_validation_iterator(iter: &mut slice::Iter<u8>)
                 }
                 _ => err!()
             }
+            offset += 1;
+        } else {
+            // Ascii case, try to skip forward quickly.
+            // When the pointer is aligned, read 2 words of data per iteration
+            // until we find a word containing a non-ascii byte.
+            const UNROLLED_BY: usize = 2;
+            let ptr = v.as_ptr();
+            let align = (ptr as usize + offset) & (usize::BYTES - 1);
+            if align == 0 {
+                if len >= UNROLLED_BY * usize::BYTES {
+                    while offset <= len - UNROLLED_BY * usize::BYTES {
+                        unsafe {
+                            let u = *(ptr.offset(offset as isize) as *const usize);
+                            let v = *(ptr.offset((offset + usize::BYTES) as isize) as *const usize);
+
+                            // break if there is a nonascii byte
+                            let zu = contains_nonascii(u);
+                            let zv = contains_nonascii(v);
+                            if zu || zv {
+                                break;
+                            }
+                        }
+                        offset += UNROLLED_BY * usize::BYTES;
```

Comment:
You could choose to introduce a constant like `const STRIDE_BYTES: usize = UNROLLED_BY * usize::BYTES` to avoid duplicating the math.